src/Resolution.ts

```diff
@@ -732,24 +772,60 @@ export default class Resolution {
     return records[newRecord] || records[oldRecord];
   }
 
-  private getNamingMethod(domain: string): NamingService | undefined {
-    return this.serviceMap[findNamingServiceName(domain)];
+  private getNamingMethods(domain: string): NamingService[] {
+    const method = this.serviceMap[findNamingServiceName(domain)];
+    if (!method) {
+      return [];
+    } else if (method === this.serviceMap.ZNS) {
+      return [this.serviceMap.UNS, method];
+    } else {
+      return [method];
+    }
   }
 
-  private getNamingMethodOrThrow(domain: string): NamingService {
-    const method = this.getNamingMethod(domain);
-    if (!method) {
+  private async callForDomain<F extends keyof NamingService>(
+    domain: string,
+    methods: NamingService[],
+    func: F,
+    args: Parameters<NamingService[F]>,
+  ): Promise<UnwrapPromise<ReturnType<NamingService[F]>>> {
+    if (!methods.length) {
       throw new ResolutionError(ResolutionErrorCode.UnsupportedDomain, {
         domain,
       });
     }
 
-    return method;
+    const servicePromises = methods.map((method) =>
+      method[func].call(method, ...args),
+    );
+
+    for (const servicePromise of servicePromises) {
+      if (servicePromise) {
+        try {
+          return await servicePromise;
+        } catch (err) {
+          if (
+            !(
+              err instanceof ResolutionError &&
```

Comment:
Ideally, I think `NamingService` should never throw `UnregisteredDomain` and it can only be thrown in this class. It will prevent code duplication as well as rethrowing. Consider implementing in case it will not take a lot to do.

Reply:
I think that it helps us to determine that a call failed due to a domain not existing in a certain chain and not due to something else. Do you suggest making every `NamingService` method involving a domain return an optional result (i.e. `T | undefined`)? I guess that can work as well, but it might be tricky to implement because `NamingService` and `Resolution` methods would have different return types.

Reply:
Yes, `T | undefined` works. If it is tricky, lets leave it to dev tools to fix later.

Reply:
I like the `T | undefined` idea too.
If it's difficult to implement now, let's add it to the dev.tools backlog.